### PR TITLE
Vimc 3197 - THIRD - touchstone_demographic_dataset support

### PR DESCRIPTION
**Pull request by weshinsley**
This adds support for touchstone_demographic_dataset.csv in the human-friendly form: `touchstone, demographic_source, demographic_statistic_type` - where all three are strings. This is stored in the db as `touchstone, demographic_dataset` where the dataset is numerical. 

We may add new `touchstone_demographic_dataset` rows, which will have a negative id as a placeholder during the update, which gets replaced with the correct serial id in the load stage. In future PRs, we will have to translate between the fake and real ids in the load stage, since we need consistency in the transform stage, but we don't know the real ids until the load stage.

https://vimc.myjetbrains.com/youtrack/issue/VIMC-3197

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -1,8 +1,8 @@
 Package: stoner
 Title: Support for building VIMC Montagu Touchstones, using dettl
-Version: 0.0.2
+Version: 0.0.3
 Authors@R:
-    c(person("Wes", "Hinsley",role = c("aut", "cre", "itr"),
+    c(person("Wes", "Hinsley",role = c("aut", "cre", "dnc", "itr"),
```
Comment (richfitz) [inline DESCRIPTION:5]: :dancer:

File: R/load.R [hunk 2]
```diff
@@ -21,4 +21,15 @@ stone_load <- function(transformed_data, con,
   load_touchstone(transformed_data, con)
   load_scenario_description(transformed_data, con,
                             allow_overwrite_scenario_description)
+  load_touchstone_demographic_dataset(transformed_data, con)
+
+  if (!is.null(con)) {
```
Comment (richfitz) [inline R/load.R]: Consider moving this into its own function: `check_faulty_serials`

File: R/touchstone_demographic_dataset.R [hunk 3]
```diff
@@ -0,0 +1,245 @@
+# Get all touchstone demographic dataset rows for a set
+# of touchstones. Also lookup both strings and numerical
+# ids for demographic source and statistic type.
+
+db_tdd_for_touchstones <- function(con, touchstones) {
+  DBI::dbGetQuery(con, sprintf("
```
Comment (richfitz) [inline R/touchstone_demographic_dataset.R:6]: would it be useful to put this (and other large SQL comments) into an external file or not?

File: R/touchstone_demographic_dataset.R [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+  # Do we have any touchstone_demographic_dataset work to do?
+
+  if (!is.null(e$touchstone_demographic_dataset_csv)) {
+
+    e_tdd_csv <- e$touchstone_demographic_dataset_csv
+
+    # Test csv columns are correct.
+
+    if (length(names(e_tdd_csv)) != 3) {
```
Comment (richfitz) [inline R/touchstone_demographic_dataset.R]: you can use `setequal(names(x), expected)` to check length and content of a group of names without regard to their order

File: R/touchstone_demographic_dataset.R [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      e_tdd_csv$touchstone, e_tdd_csv$demographic_source,
+      e_tdd_csv$demographic_statistic_type, sep = '\r')
+
+    e$db_tdd$mash <- paste(e$db_tdd$touchstone,
+                           e$db_tdd$dsource_code,
+                           e$db_tdd$dtype_code, sep = '\r')
+
+    # A mash to compare with existing demographic_datasets
+
+    e$touchstone_demographic_dataset_csv$ds_mash <- paste(
```
Comment (richfitz) [inline R/touchstone_demographic_dataset.R]: do we not have a `mash` function we should be using here?

File: R/touchstone_demographic_dataset.R [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+###############################################################################
+
+transform_touchstone_demographic_dataset <- function(e) {
+
+  tdd <- e$touchstone_demographic_dataset_csv
+
+  if (!is.null(tdd)) {
+    tdd$demographic_dataset <- e$db_dd$id[match(tdd$ds_mash, e$db_dd$mash)]
+
+  # If any (touchstone,dataset) already exist in db's
```
Comment (richfitz) [inline R/touchstone_demographic_dataset.R:162]: indenting needs redoing

File: R/touchstone_demographic_dataset.R [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      tdd$id <- e$db_tdd$id[match(tdd$mash, e$db_tdd$mash)]
+
+      which_nas <- which(is.na(tdd$id))
+      tdd$id[which_nas] <- seq(from = -1, by = -1,
+                               length.out = length(which_nas))
+
+      tdd$already_exists_db <- FALSE
+      tdd <- tdd[, c("already_exists_db", "demographic_dataset",
+                     "id", "touchstone")]
+      list(touchstone_demographic_dataset = tdd)
```
Comment (richfitz) [inline R/touchstone_demographic_dataset.R]: you might find the early-exit/happy path optimising approach nicer to use here

**Review by richfitz — CHANGES_REQUESTED** [review]
Minor things but otherwise looking good

File: DESCRIPTION (hunk 1 above)
Comment (weshinsley) [reply DESCRIPTION:5]: This can only get more exciting the more stoner PRs we get through.

File: R/touchstone_demographic_dataset.R (hunk 3 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R:6]: I think they'll be fairly specific to their context - these won't get any use outside of the touchstone_demographic_dataset world (I think). Will make a note to review this in a couple of PRs where I might have some more SQL to ponder. It does make the .R file a bit clumpy...

File: R/touchstone_demographic_dataset.R (hunk 5 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R]: My mash function in utils.R (at the moment) does all fields in a generic unnamed way; this one mashes a named subset. I could parameterise the existing mash to optionally take names I guess...

File: R/touchstone_demographic_dataset.R (hunk 6 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R:162]: Done - as product of the below change...

File: R/touchstone_demographic_dataset.R (hunk 7 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R]: Done - yes, much nicer.

File: R/touchstone_demographic_dataset.R (hunk 4 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R]: Done - nice.

File: R/touchstone_demographic_dataset.R (hunk 7 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R]: Applied this a bit earlier in the extract too.

File: R/load.R (hunk 2 above)
Comment (weshinsley) [reply R/load.R]: Have merged it into one function in utils.

File: R/touchstone_demographic_dataset.R (hunk 5 above)
Comment (weshinsley) [reply R/touchstone_demographic_dataset.R]: Done - and discovered the first line of old mash function was doing nothing useful.
Also added a couple of tests for mash in test_utils.R

**Review by richfitz — APPROVED** [review]

**Pull request merged**